### PR TITLE
Add #[track_caller] to panicking Vec functions

### PR DESCRIPTION
**Pull request by Dylan-DPC-zz**
Rework of #79323

cc @nvzqz 

r? @joshtriplett

### CODE REVIEW

**Comment by nagisa** [comment]
This definitely needs a performance evaluation, in both microbenchmark and real use-case scenarios. `#[track_caller]` may be cheap, but its not free.

**Comment by joshtriplett** [comment]
There's a minor performance hit by timing metrics, but the hit on memory (max-rss) seems more excessive.

File: library/alloc/src/vec/mod.rs [hunk 1]
```diff
@@ -2792,7 +2807,7 @@ impl<T, A: Allocator, const N: usize> TryFrom<Vec<T, A>> for [T; N] {
     /// # Examples
     ///
     /// ```
-    /// use std::convert::TryInto;
+    /// use std::convert::TryInto;k
```
Comment (Thomasdezeeuw) [inline library/alloc/src/vec/mod.rs:2810]: This seems unintentional.

**Comment by joshtriplett** [comment]
Following up on this: I've since learned that the `max-rss` statistics aren't necessarily reliable enough to know if the observed increase in memory here is real or spurious. The performance numbers seem reasonable given the benefit, so I think it's reasonable for this change to go ahead (modulo the typo fix already noted).

**Comment by Dylan-DPC** [comment]
@joshtriplett once i fix the typo, can i r= you?

**Comment by joshtriplett** [comment]
@Dylan-DPC Yes, go ahead.

**Comment by Dylan-DPC** [comment]
created #83909 because i did this on the wrong remote :D

**Pull request closed without merging**